### PR TITLE
[rfxcom] Support for RFXSensor

### PR DESCRIPTION
**Pull request by martinvw**
See related issue https://github.com/openhab/openhab2-addons/issues/1934

I never received the feedback so I still have this stall branch

### CODE REVIEW

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComRFXSensorMessage.java [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    }
+
+    private void encodeVoltageMessage(byte msg1, byte msg2) {
+        miliVoltageTimesTen = BigDecimal.valueOf((short) ((msg1 & 0xFF) << 8 | (msg2 & 0xFF)));
+    }
+
+    private void encodeStatusMessage(byte msg2) {
+
+
+        //logger
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComRFXSensorMessage.java]: Note to self, what is this?

**Comment by martinvw** [comment]
It would be nice to have someone around to test this change, volunteers? 

A testing JAR could be found in the issue #1934

File: bundles/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComRFXSensorMessage.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        if (temperature < 0) {
+            data[5] |= 0x80;
+        }
+    }
+
+    private void decodeVoltageMessage(byte[] data) {
+        short miliVoltageTimesTenShort = this.miliVoltageTimesTen.shortValueExact();
+        data[5] = (byte) ((miliVoltageTimesTenShort >> 8) & 0xFF);
+        data[6] = (byte) (miliVoltageTimesTenShort & 0xFF);
+
```
Comment (martinvw) [inline bundles/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComRFXSensorMessage.java]: Remove newline

File: bundles/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComRFXSensorMessage.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+            temperature = -temperature;
+        }
+    }
+
+    private void encodeVoltageMessage(byte msg1, byte msg2) {
+        miliVoltageTimesTen = BigDecimal.valueOf((short) ((msg1 & 0xFF) << 8 | (msg2 & 0xFF)));
+    }
+
+    private void encodeStatusMessage(byte msg2) {
+        //logger
```
Comment (martinvw) [inline bundles/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComRFXSensorMessage.java]: There is a TODO, what should be done and can it be completed before merging the PR?

File: bundles/org.openhab.binding.rfxcom/src/main/resources/ESH-INF/thing/rfxsensor.xml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+			<bridge-type-ref id="RFXtrx433" />
+			<bridge-type-ref id="RFXrec433" />
+		</supported-bridge-type-refs>
+
+		<label>RFXCOM Temperature Sensor</label>
+		<description>A Temperature device.</description>
+
+		<channels>
+			<channel id="temperature" typeId="temperature" />
+			<!-- TODO add two more channel here -->
```
Comment (martinvw) [inline bundles/org.openhab.binding.rfxcom/src/main/resources/ESH-INF/thing/rfxsensor.xml]: There is a TODO, what should be done and can it be completed before merging the PR?

File: bundles/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComRFXSensorMessage.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public State convertToState(String channelId) throws RFXComUnsupportedChannelException {
+        switch (channelId) {
+            case CHANNEL_TEMPERATURE:
+                return temperature == null ? UnDefType.UNDEF : new DecimalType(temperature);
+
+            case CHANNEL_VOLTAGE:
```
Comment (martinvw) [inline bundles/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComRFXSensorMessage.java]: Add this missing channel

**Comment by J-N-K** [comment]
Is this still work in progress?

**Comment by martinvw** [comment]
Yes, I need to build the cache, to do some calculations:

https://github.com/openhab/openhab2-addons/issues/1934#issuecomment-504883966

But also have some work for the RFXCOM fan :-)

**Comment by martinvw** [comment]
Note to self: the readme has net yet been updated

**Comment by martinvw** [comment]
Readme also done.

**Comment by kaikreuzer** [comment]
That's a big one - who'd be the best to review (and test)? @paulianttila?

**Comment by martinvw** [comment]
> That's a big one - who'd be the best to review (and test)? @paulianttila?

That would indeed be great.

However, a lot of the changes are just repeated to the other files, so the total change is not that big. I would think that some of the other usuals like @Jamstah could maybe also review it if @paulianttila does not have time.

I'm running it in my production system currently I did not spot any unexpected behaviour, I will also in the original issue for testers.

File: bundles/org.openhab.binding.rfxcom/src/main/resources/ESH-INF/thing/rfxsensor.xml [hunk 6]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<thing:thing-descriptions bindingId="rfxcom" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.rfxcom/src/main/resources/ESH-INF/thing/rfxsensor.xml]: Schema location needs update. Maybe rebase to get 0.8.0 SAT version and rerun to check xml files.

Comment (martinvw) [reply bundles/org.openhab.binding.rfxcom/src/main/resources/ESH-INF/thing/rfxsensor.xml]: The good thing is that I addressed this and the other thing is that I addressed some of the SAT findings. @Hilbrand would be great if you could do the first review. Thanks!

File: bundles/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComRFXSensorMessage.java [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+
+    private void decodeVoltageMessage(byte[] data) {
+        short miliVoltageTimesTenShort = this.miliVoltageTimesTen.shortValueExact();
+        data[5] = (byte) ((miliVoltageTimesTenShort >> 8) & 0xFF);
+        data[6] = (byte) (miliVoltageTimesTenShort & 0xFF);
+
+    }
+
+    private void decodeStatusMessage(byte[] data) {
+        logger.info("A status message was received {}", HexUtils.bytesToHex(data));
```
[the remark this answers is not in the document]
Comment (martinvw) [reply bundles/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComRFXSensorMessage.java]: I tried to get my hands on this output while testing with a user but he never received so I'm still interested in that input.

File: bundles/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComRFXSensorMessage.java [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+            default:
+                return super.convertToState(channelId, deviceState);
+        }
+    }
+
+    private State handleVoltage() {
+        if (miliVoltageTimesTen != null) {
+            return new DecimalType(getVoltage());
+        } else {
+            return UnDefType.UNDEF;
```
[the remark this answers is not in the document]
Comment (martinvw) [reply bundles/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComRFXSensorMessage.java]: That is a good question, let me think about this

File: bundles/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComRFXSensorMessage.java [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+        }
+    }
+
+    private State handleHumidity(DeviceState deviceState) {
+        DecimalType temperatureState = (DecimalType) deviceState.getLastState(CHANNEL_TEMPERATURE);
+        DecimalType referenceVoltageState = (DecimalType) deviceState.getLastState(CHANNEL_REFERENCE_VOLTAGE);
```
[the remark this answers is not in the document]
Comment (martinvw) [reply bundles/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComRFXSensorMessage.java]: Nice catch 👍

File: bundles/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComRFXSensorMessage.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (martinvw) [reply bundles/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComRFXSensorMessage.java]: It's correct, although it makes the class-cast problem even bigger :-)

Returning `null` means (in this context) that there is no value and that we should not even update the core about the value, but if we receive a message of the correct sub-type and it does not contain a voltage we should send an UNDEF to the core.

I also added a `getTemperature` to make this really consistent. And I handled the potential class-cast.

**Comment by martinvw** [comment]
@Hilbrand thanks for your comments, they should all be processed now.

**Comment by Hilbrand** [comment]
@martinvw Do you still want to have some feedback of the other testers? Otherwise I'll merge this pr.

**Comment by martinvw** [comment]
If there is any feedback feel free to still shout out, I can most likely pick it up rather quickly :-)

**Pull request merged**